ValueMapping: Added support for mapping text to color, boolean values, NaN and Null. Improved UI for value mapping.

torkelo opened the pull request:
Closes #33755
Closes #26837

* [x] Update definition model to include maps and ranges 
* [x] Migration to new model
* [x] Migration for value mappings in override rules 
* [x] New model no longer maps text to numbers (that are later processed with thresholds) 
* [x] Update view only editor (sidebar) UI 
* [x] New modal base edit table U
* [x] Add support for range mappings 
* [x] Use new color option in display processor 
* [x] New NullToText map type to map null, undefined an NaN 

WIP UI 

![Screenshot from 2021-05-10 16-55-32](https://user-images.githubusercontent.com/10999/117679627-bc831600-b1b0-11eb-8401-039c8e4ff03e.png)

ryantxu committed:
alternative mapping editor

ryantxu committed:
alternative mapping editor

ryantxu committed:
values updating

dprokop committed:
UI updates

ryantxu committed:
merge master

ryantxu committed:
remove empty operators

ryantxu committed:
fix types

ryantxu committed:
horizontal

ryantxu committed:
merge master

ryantxu committed:
Merge remote-tracking branch 'origin/master' into mapping-as-string

ryantxu committed:
merge master

torkelo committed:
This is tricky

torkelo committed:
Merge remote-tracking branch 'origin/main' into value-mapping-ui-v2

torkelo committed:
New value mapping model and migration

ryantxu committed:
DataSource: show the uid in edit url, not the local id (#33818)

ryantxu committed:
update mapping model object

ryantxu committed:
Merge remote-tracking branch 'origin/main' into value-mapping-model2

torkelo committed:
Update to UI

torkelo committed:
fixing ts issues

torkelo committed:
Merge branch 'value-mapping-model2' of github.com:grafana/grafana into value-mapping-model2

torkelo committed:
Editing starting to work

torkelo committed:
adding missing thing

torkelo committed:
Update display processor to use color from value mapping

ryantxu commented on packages/grafana-data/src/field/displayProcessor.ts:
```diff
@@ -62,20 +64,28 @@ export function getDisplayProcessor(options?: DisplayProcessorOptions): DisplayP
-      if (mappedValue) {
-        text = mappedValue.text;
-        const v = isStringUnit ? NaN : toNumber(text);
+        if (mappingResult.value !== undefined && !isStringUnit) {
+          numeric = mappingResult.value;
+        }
```
Is this really necessary?  We could remove the number option entirely now -- that can be supported via transformations if it is really required.  We always do `shouldFormat = false;` so the unit does not matter

torkelo committed:
Range maps now work

torkelo committed:
Working on unit tests for modal editor

torkelo committed:
Updated

torkelo committed:
Merge remote-tracking branch 'origin/main' into value-mapping-model2

papagian committed:
[Alerting]: store encrypted receiver secure settings (#33832)

* [Alerting]: Store secure settings encrypted

* Move encryption to the API handler

sunker committed:
CloudMonitoring: Migrate config editor from angular to react (#33645)

* fix broken config ctrl

* replace angular config with react config editor

* remove not used code

* add extra linebreak

* add noopener to link

* only test jwt props that we actually need

Elfo404 committed:
Elasticsearch: automatically set date_histogram field based on data source configuration (#33840)

danieldspx committed:
Docs: delete from high availability docs references to removed configurations related to session storage (#33827)

* docs: delete from high availability docs references to removed configurations related to session storage

* docs: remove session storage mention and focus on the auth token implementation

torkelo committed:
Adding new NullToText mapping type

ying-jeanne committed:
fix postgres to have precision of ms (#33853)

Clarity-89 committed:
Use ids for enterprise nav model items (#33854)

kylebrandt committed:
Alerting: Disable dash alerting if NG enabled (#33794)

dsotirakis committed:
Scuemata: Add grafana-cli cue schema validation to CI (#33798)

* Add scuemata validation in CI

* Fixes according to reviewer's comments

torkelo committed:
Added null to text UI

wbrowne committed:
Ensure http client has no timeout (#33856)

zserge committed:
Redact sensitive values before logging them (#33829)

* use a common way to redact sensitive values before logging them

* fix panic on missing testCase.err, simplify require checks

* fix a silly typo

* combine readConfig and buildConnectionString methods, as they are closely related

davkal committed:
Tempo: Search for Traces by querying Loki directly from Tempo (#33308)

* Loki query from Tempo UI

- add query type selector to tempo
- introduce linkedDatasource concept that runs queries on behalf of another datasource
- Tempo uses Loki's query field and Loki's derived fields to find a trace matcher
- Tempo uses the trace-to-logs mechanism to determine which dataource is linked

Loki data loads successfully via tempo

Extracted result transformers

Skip null values

Show trace on list id click

Query type selector

Use linked field trace regexp

* Review feedback

ryantxu committed:
add color from old threshold config

zserge committed:
Add isolation level db configuration parameter (#33830)

* add isolation level db configuration parameter

* add isolation_level to default.ini and sample.ini

* add note that only mysql supports isolation levels for now

* mention isolation_level in the documentation

* Update docs/sources/administration/configuration.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

thisisobate committed:
Drawer: fixes title overflowing its container (#33857)

ryantxu committed:
Timeline: move grafana/ui elements to the panel folder (#33803)

owen-d committed:
revendor loki with new Tripperware (#33858)

FZambia committed:
live: move connection endpoint to api scope, fixes #33861 (#33863)

jvoeller committed:
OAuth: Add support for empty scopes (#32129)

* add parameter empty_scopes to override scope parameter with empty value and thus be able to authenticate against IdPs without scopes. Issue #27503

Update docs/sources/auth/generic-oauth.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

* updated check according to feedback

* Update generic-oauth.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

zoltanbedi committed:
Prometheus: Fix exemplars hover disappearing and broken link (#33866)

* Revert "Tooltip: eliminate flickering when repaint can't keep up (#33609)"

This reverts commit e159985aa2907e2c2889853f9295183edc1032ac.

* Fix exemplar linking

Co-authored-by: David Kaltschmidt <david.kaltschmidt@gmail.com>

achatterjee-grafana committed:
Removed content as per MarcusE's suggestion in https://github.com/grafana/grafana/issues/33822. (#33870)

achatterjee-grafana committed:
Fixed grammar usage. (#33871)

aocenas committed:
Explore: Wrap each panel in separate error boundary (#33868)

leeoniya committed:
New Panel: Histogram (#33752)

ivanahuckova committed:
Sanitize PromLink button (#33874)

torkelo committed:
Added migration for overrides, added Type column

torkelo committed:
Merge branch 'value-mapping-model2' of github.com:grafana/grafana into value-mapping-model2

oscarkilhed committed:
Refactor and unify option creation between new visualizations (#33867)

* Refactor and unify option creation between new visualizations

* move to grafana/ui

* move to grafana/ui

* resolve duplicate scale config

* more imports

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

ryantxu committed:
Live: do not show connection warning when on the login page (#33865)

torkelo committed:
Added compact view model with color edit capability

owen-d committed:
enforce receivers align with backend type when posting AM config (#33877)

ryantxu committed:
special values

ryantxu committed:
Merge remote-tracking branch 'origin/main' into value-mapping-model2

ryantxu commented:
@torkelo I think you are missing a file -- I don't see the compact view in history :man_shrugging: 

I changed the "NullMode" to "Special" and added true/false -- the dropdown needs some style help!  
![image](https://user-images.githubusercontent.com/705951/117729414-a456cb00-b19f-11eb-8705-e5d04350aad9.png)

Tristan971 committed:
Document `hide_version` flag (#33670)

Unauthenticated users can be barred from being shown the current Grafana server version since https://github.com/grafana/grafana/pull/24919

leeoniya committed:
GraphNG: always use "x" as scaleKey for x axis (#33884)

leeoniya committed:
Timeline: add support for strings & booleans (#33882)

dependabot[bot] committed:
Chore(deps): Bump hosted-git-info from 2.8.5 to 2.8.9 (#33886)

Bumps [hosted-git-info](https://github.com/npm/hosted-git-info) from 2.8.5 to 2.8.9.
- [Release notes](https://github.com/npm/hosted-git-info/releases)
- [Changelog](https://github.com/npm/hosted-git-info/blob/v2.8.9/CHANGELOG.md)
- [Commits](https://github.com/npm/hosted-git-info/compare/v2.8.5...v2.8.9)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

torkelo committed:
Merge branch 'value-mapping-model2' of github.com:grafana/grafana into value-mapping-model2

torkelo committed:
merge fix

ryantxu committed:
merge with torkel

ryantxu committed:
add empty special character

torkelo committed:
Fixed centered text in special value match select

torkelo committed:
fixed unit tests

torkelo committed:
Updated snapshot

torkelo committed:
Update dashboard page

torkelo committed:
updated snapshot

torkelo committed:
Fix more unit tests

torkelo committed:
Fixed test

torkelo committed:
Merge remote-tracking branch 'origin' into value-mapping-model2

torkelo committed:
Updates

torkelo committed:
Added back tests

hugohaggmark commented on packages/grafana-data/src/utils/anyToNumber.ts:
```diff
@@ -0,0 +1,18 @@
+    return value ? 1 : 0;
+  }
+
+  return toNumber(value);
```
nit: Could we use Number instead?
```suggestion
  return Number(value);
```

torkelo replied:
not sure, this was only copied code. Think there are some differences with toNumber vs Number constructor

hugohaggmark commented on packages/grafana-data/src/utils/valueMappings.ts:
```diff
@@ -1,104 +1,88 @@
-      return allValueMappings.concat(valueToTextMapping);
+import { ValueMapping, MappingType, ValueMappingResult, SpecialValueMatch } from '../types';
+
+export function getValueMappingResult(valueMappings: ValueMapping[], value: any): ValueMappingResult | null {
```
Would a FieldRegistry for ValueMappings be a nicer approach?

torkelo replied:
maybe, but do no foresee many of these mapping types, think these 3 will suffice (famous last words).

hugohaggmark commented on public/app/plugins/panel/timeline/utils.ts:
```diff
@@ -74,8 +73,7 @@ export const preparePlotConfigBuilder: PrepConfig = ({
     const field = frame.fields[seriesIdx];
-    const mode = field.config?.color?.mode;
-    if (mode && field.display && (mode === FieldColorModeId.Thresholds || mode.startsWith('continuous-'))) {
+    if (field.display) {
```
Hmm, how does this change relate to value mapping?

hugohaggmark reviewed (approved):
Looks great, going to take it for a test spin after this!

hugohaggmark reviewed (commented):
Tested it with Postgres with different scenarios and works as expected!

jackw reviewed (approved):
I'm not all that familiar with the feature but tested it locally with stat panel and null, range and value mappings and works as expected. 🚀

torkelo commented on packages/grafana-ui/src/components/ValueMappingsEditor/ValueMappingsEditor.tsx:
```diff
@@ -1,60 +1,90 @@
+
+export const getStyles = (theme: GrafanaTheme2) => ({
+  modal: css({
+    width: '980px',
```
there is a max-width: 100% on the default modal styles that makes this work / adapt even to smaller screens

torkelo committed:
Fixed doc issue

torkelo committed:
Merge remote-tracking branch 'origin/main' into value-mapping-model2

leeoniya commented:
nice.

oddlittlebird commented:
@torkelo, are you planning to write docs for this?

torkelo reviewed (commented):
The docs should already be updated as part of this PR. But only bare minimum update. Think the docs is missing the special type, that was added after the docs update and missed that.

oddlittlebird commented:
Ah, I see
